# Practice Case 1

Author: Gilang Mahardika
<br> Date: 24 October 2019

## Part A
On your Jupyter notebook, answer the following questions!
1. How many rows and columns reflected in the dataset?
2. What is the minimum, maximum, and mean of the customer's age?
3. How many are male and how many are female? Why?
4. How many high skilled workers customers reflected in the dataset?
5. How many customers who own the house?
6. How many customers who have 'rich' Saving accounts?
7. How many customers who do not have checking accounts or have zero checking account?
8. Who is the one who have highest credit amount? Show us the profile of the age, sex, job, housing, saving, account, and checking account!
9. How many customers are classified as good risk and bad risk?
10. Is the average of credit amount for good risk is higher than bad risk?

### Import Packages

In [1]:
import pandas as pd
import numpy as np

### Import Dataset

In [2]:
df = pd.read_csv('german_credit_data_with_risk.csv', encoding='latin1')
df.head()

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0   67    male    2     own             NaN           little           1169   
1   22  female    2     own          little         moderate           5951   
2   49    male    1     own          little              NaN           2096   
3   45    male    2    free          little           little           7882   
4   53    male    2    free          little           little           4870   

   Duration              Purpose  Risk  
0         6             radio/TV  good  
1        48             radio/TV   bad  
2        12            education  good  
3        42  furniture/equipment  good  
4        24                  car   bad

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
Age                 1000 non-null int64
Sex                 1000 non-null object
Job                 1000 non-null int64
Housing             1000 non-null object
Saving accounts     817 non-null object
Checking account    606 non-null object
Credit amount       1000 non-null int64
Duration            1000 non-null int64
Purpose             1000 non-null object
Risk                1000 non-null object
dtypes: int64(4), object(6)
memory usage: 78.2+ KB


In [4]:
df['Job'] = df['Job'].astype('object')
df['Duration'] = df['Duration'].astype('object')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
Age                 1000 non-null int64
Sex                 1000 non-null object
Job                 1000 non-null object
Housing             1000 non-null object
Saving accounts     817 non-null object
Checking account    606 non-null object
Credit amount       1000 non-null int64
Duration            1000 non-null object
Purpose             1000 non-null object
Risk                1000 non-null object
dtypes: int64(2), object(8)
memory usage: 78.2+ KB


**1. How many rows and columns reflected in the dataset?**

In [6]:
df.shape

(1000, 10)

**2. What is the minimum, maximum, and mean of the customer's age**

In [7]:
df.Age.describe()[['min', 'max', 'mean']]

min     19.000
max     75.000
mean    35.546
Name: Age, dtype: float64

**3. How many are male and how many are female? Why?**

In [8]:
df['Sex'].value_counts()

male      690
female    310
Name: Sex, dtype: int64

It may be men willing to take more loan, but this also culd not be the case, since there are also other reason, such as regulation

**4. How many high skilled workers customers reflected in the dataset?**

In [9]:
df[df['Job'] == 3][['Job']].count()

Job    148
dtype: int64

**5. How many customers who own the house?**

In [10]:
df[df['Housing'] == 'own'][['Housing']].count()

Housing    713
dtype: int64

**6. How many customers who have 'rich' Saving accounts?**

In [11]:
df[df['Saving accounts'] == 'rich'][['Saving accounts']].count()

Saving accounts    48
dtype: int64

**7. How many customers who do not have checking accounts or have zero checking account?**

In [12]:
df[['Checking account']].isnull().sum()

Checking account    394
dtype: int64

With assumption that **NaN** is zero or don't have checking account

**8. Who is the one who have highest credit amount? Show us the profile of the age, sex, job, housing, saving account, and checking account!**

In [13]:
df[['Age','Sex', 'Job','Housing', 'Saving accounts', 'Checking account', 'Credit amount']].sort_values(by=['Credit amount'], ascending=False).head(1)

Age     Sex Job Housing Saving accounts Checking account  Credit amount
915   32  female   3     own          little         moderate          18424

**9. How many customers are classified as good risk and bad risk?** 

In [14]:
df['Risk'].value_counts()

good    700
bad     300
Name: Risk, dtype: int64

**10. Is the average of credit amount for good risk is higher than bad risk?**

In [15]:
print(df[df['Risk'] == 'good']['Credit amount'].mean() > df[df['Risk'] == 'bad']['Credit amount'].mean())

False


Which means that, in average, credit amount with bad risk is higher than the good risk in other words, the data tends to accept higher credit amount but also with bad risk

## Part B
Aside questions mentioned on Part A, create 10 more questions and answers to understand more about the dataset. 

**1. What is the characteristics category of purpose financing that was applied by customers**

In [16]:
df['Purpose'].value_counts()

car                    337
radio/TV               280
furniture/equipment    181
business                97
education               59
repairs                 22
vacation/others         12
domestic appliances     12
Name: Purpose, dtype: int64

**2. What is the characteristic of loan duration of the customers?**

In [17]:
df['Duration'].value_counts().head()

24    184
12    179
18    113
36     83
6      75
Name: Duration, dtype: int64

It seems that our client have duration mostly at 24 months

**3. How is the characteristics betwen credit amount and job of our clients?**

In [18]:
df.groupby(['Job'])[['Credit amount']].agg(np.mean)

Credit amount
Job               
0      2745.136364
1      2358.520000
2      3070.965079
3      5435.493243

It seems that clients with highly skilled job have higher average credit amount around 5,435 euro

**4. How is the characteristics betwen credit amount and loan purpose of our clients?**

In [19]:
df.groupby(['Purpose'])[['Credit amount']].agg([np.sum, np.mean, np.median])

Credit amount                    
                              sum         mean median
Purpose                                              
business                   403330  4158.041237   3161
car                       1269881  3768.192878   2679
domestic appliances         17976  1498.000000   1249
education                  169873  2879.203390   1597
furniture/equipment        555125  3066.988950   2578
radio/TV                   696543  2487.653571   1890
repairs                     60018  2728.090909   1749
vacation/others             98512  8209.333333   6948

As the highest average of the amount credit is vacation/others this because of small applicants in vacation/others compare to car. But if we look at the total amount credit of purpose car is the highest total amount credit due to high number of applicants 

**5. How is the characteristics betwen risk and gender of our clients?**

In [20]:
pd.crosstab(df['Sex'], df['Risk'], normalize='index')

Risk         bad      good
Sex                       
female  0.351613  0.648387
male    0.276812  0.723188

**6. How is the characteristics betwen risk and purpose of our clients?**

In [21]:
pd.crosstab(df['Purpose'], df['Risk'], normalize='index')

Risk                      bad      good
Purpose                                
business             0.350515  0.649485
car                  0.314540  0.685460
domestic appliances  0.333333  0.666667
education            0.389831  0.610169
furniture/equipment  0.320442  0.679558
radio/TV             0.221429  0.778571
repairs              0.363636  0.636364
vacation/others      0.416667  0.583333

**7. How is the characteristics betwen risk and house of our clients?**

In [22]:
pd.crosstab(df['Housing'], df['Risk'], normalize='index')

Risk          bad      good
Housing                    
free     0.407407  0.592593
own      0.260870  0.739130
rent     0.391061  0.608939

**8. How is the characteristics betwen risk and Saving Accounts of our clients?**

In [23]:
pd.crosstab(df['Saving accounts'], df['Risk'], normalize='index')

Risk                  bad      good
Saving accounts                    
little           0.359867  0.640133
moderate         0.330097  0.669903
quite rich       0.174603  0.825397
rich             0.125000  0.875000

**9. How is the characteristics betwen risk and Saving Accounts of our clients?**

In [24]:
pd.crosstab(df['Checking account'], df['Risk'], normalize='index')

Risk                   bad      good
Checking account                    
little            0.492701  0.507299
moderate          0.390335  0.609665
rich              0.222222  0.777778

**10. How is the characteristics betwen risk and Duration of our clients?**

In [28]:
pd.crosstab(df['Duration'], df['Risk'])

Risk      bad  good
Duration           
4           0     6
5           0     1
6           9    66
7           0     5
8           1     6
9          14    35
10          3    25
11          0     9
12         49   130
13          0     4
14          1     3
15         12    52
16          1     1
18         42    71
20          1     7
21          9    21
22          0     2
24         56   128
26          0     1
27          5     8
28          1     2
30         13    27
33          1     2
36         37    46
39          1     4
40          1     0
42          3     8
45          4     1
47          0     1
48         28    20
54          1     1
60          6     7
72          1     0

## Part C

Define at least 3 business problems that you can find based on the dataset!

* Don't have income data, I think it's better to see the relationships between income and risk of each person

* It seems that duration more than 24 months has roughly bad risk number and good risk number

* Streghten loan on Radio or Tv, since it shows low bad rate and also good total amount of credit